java.lang.IllegalArgumentException: from and to edge cannot be NO_EDGE

Problem deploying osm.pbf-files to ors in docker/kubernetes

#### Here's what I did
Deployed docker image on kubernetes using both latest and v6.4.1 tag.
Provided norway-latest.osm.pbf and phillippines-latest.osm.pbf from Geofabrik.

---
#### Here's what I got
java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: from and to edge cannot be NO_EDGE
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.heigit.ors.routing.RoutingProfileManager.initialize(RoutingProfileManager.java:167) ~[classes/:6.4.1]
	at org.heigit.ors.routing.RoutingProfileManager.getInstance(RoutingProfileManager.java:62) ~[classes/:6.4.1]
	at org.heigit.ors.servlet.listeners.ORSInitContextListener.lambda$contextInitialized$0(ORSInitContextListener.java:40) ~[classes/:6.4.1]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: from and to edge cannot be NO_EDGE
	at org.heigit.ors.routing.graphhopper.extensions.core.CoreAlgoFactoryDecorator.prepare(CoreAlgoFactoryDecorator.java:306) ~[classes/:6.4.1]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.prepareCore(ORSGraphHopper.java:737) ~[classes/:6.4.1]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.postProcessing(ORSGraphHopper.java:657) ~[classes/:6.4.1]
	at com.graphhopper.GraphHopper.process(GraphHopper.java:703) ~[graphhopper-core-v0.13.10.jar:?]
	at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:680) ~[graphhopper-core-v0.13.10.jar:?]
	at org.heigit.ors.routing.graphhopper.extensions.ORSGraphHopper.importOrLoad(ORSGraphHopper.java:155) ~[classes/:6.4.1]
	at org.heigit.ors.routing.RoutingProfile.initGraphHopper(RoutingProfile.java:194) ~[classes/:6.4.1]
	at org.heigit.ors.routing.RoutingProfile.<init>(RoutingProfile.java:138) ~[classes/:6.4.1]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:35) ~[classes/:6.4.1]
	at org.heigit.ors.routing.RoutingProfileLoader.call(RoutingProfileLoader.java:21) ~[classes/:6.4.1]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	... 1 more

---

#### Dockerfile
FROM openrouteservice/openrouteservice:v6.4.1

ENV BUILD_GRAPHS=False
ENV JAVA_OPTS="-Djava.awt.headless=true -server -XX:TargetSurvivorRatio=75 -XX:SurvivorRatio=64 -XX:MaxTenuringThreshold=3 -XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=4 -Xms8g -Xmx8g"
ENV CATALINA_OPTS="-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=9001 -Dcom.sun.management.jmxremote.rmi.port=9001 -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Djava.rmi.server.hostname=localhost"

COPY app.config /ors-core/openrouteservice/src/main/resources/
COPY custom-entrypoint.sh /

RUN apt-get -y install awscli

RUN aws s3 cp s3:/xxx.xxx.xxx/data3.osm.pbf /ors-core/data/data.osm.pbf

ENTRYPOINT []
CMD ["/bin/bash", "/custom-entrypoint.sh"]

---

#### custom-entrypoint.sh
rm -rf /ors-core/data/graphs/car
rm -rf /ors-core/data/elevation_cache/*

/docker-entrypoint.sh

---

Have anyone experienced this problem?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,8 +38,9 @@ RELEASING:
 - Allow to disable OSM conditional access and speed encoders via parameter in config file
 - Turkish language support (thanks to [kucar17](https://github.com/kucar17) for the translation)
 ### Fixed
-- Fixed isochrones algorithm selection for location_type parameter ([#676](https://github.com/GIScience/openrouteservice/issues/676)
+- Fixed isochrones algorithm selection for location_type parameter ([#676](https://github.com/GIScience/openrouteservice/issues/676))
 - Updated link to client translations in readme
+- Concurrency bug in core edge filters which caused crashes during CALT graph preparation ([#905](https://github.com/GIScience/openrouteservice/issues/905))
 
 ## [6.4.1] - 2021-03-31
 ### Fixed

--- a/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java
+++ b/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java
@@ -51,14 +51,11 @@ import org.heigit.ors.fastisochrones.partitioning.FastIsochroneFactory;
 import org.heigit.ors.routing.AvoidFeatureFlags;
 import org.heigit.ors.routing.RouteSearchContext;
 import org.heigit.ors.routing.RouteSearchParameters;
-import org.heigit.ors.routing.RoutingProfileCategory;
 import org.heigit.ors.routing.graphhopper.extensions.core.CoreAlgoFactoryDecorator;
 import org.heigit.ors.routing.graphhopper.extensions.core.CoreLMAlgoFactoryDecorator;
 import org.heigit.ors.routing.graphhopper.extensions.core.PrepareCore;
 import org.heigit.ors.routing.graphhopper.extensions.edgefilters.AvoidFeaturesEdgeFilter;
 import org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence;
-import org.heigit.ors.routing.graphhopper.extensions.edgefilters.core.*;
-import org.heigit.ors.routing.graphhopper.extensions.flagencoders.FlagEncoderNames;
 import org.heigit.ors.routing.graphhopper.extensions.storages.BordersGraphStorage;
 import org.heigit.ors.routing.graphhopper.extensions.storages.GraphStorageUtils;
 import org.heigit.ors.routing.graphhopper.extensions.util.ORSPMap;
@@ -605,54 +602,9 @@ public class ORSGraphHopper extends GraphHopper {
 
 		GraphHopperStorage gs = getGraphHopperStorage();
 
-		EncodingManager encodingManager = getEncodingManager();
-
-		int routingProfileCategory = RoutingProfileCategory.getFromEncoder(encodingManager);
-
-		/* Initialize edge filter sequence */
-
-		EdgeFilterSequence coreEdgeFilter = new EdgeFilterSequence();
-		/* Heavy vehicle filter */
-
-		if (encodingManager.hasEncoder(FlagEncoderNames.HEAVYVEHICLE)) {
-			coreEdgeFilter.add(new HeavyVehicleCoreEdgeFilter(gs));
-		}
-
-		/* Avoid features */
-
-		if ((routingProfileCategory & (RoutingProfileCategory.DRIVING | RoutingProfileCategory.CYCLING | RoutingProfileCategory.WALKING | RoutingProfileCategory.WHEELCHAIR)) != 0) {
-			coreEdgeFilter.add(new AvoidFeaturesCoreEdgeFilter(gs, routingProfileCategory));
-		}
-
-		/* Avoid borders of some form */
-
-		if ((routingProfileCategory & (RoutingProfileCategory.DRIVING | RoutingProfileCategory.CYCLING)) != 0) {
-			coreEdgeFilter.add(new AvoidBordersCoreEdgeFilter(gs));
-		}
-
-		if (routingProfileCategory == RoutingProfileCategory.WHEELCHAIR) {
-			coreEdgeFilter.add(new WheelchairCoreEdgeFilter(gs));
-		}
-
-		/* Maximum Speed Filter & Turn Restrictions */
-		if ((routingProfileCategory & RoutingProfileCategory.DRIVING) !=0 ) {
-			String[] encoders = {FlagEncoderNames.CAR_ORS, FlagEncoderNames.HEAVYVEHICLE};
-			for (String encoderName: encoders) {
-				if (encodingManager.hasEncoder(encoderName)) {
-					FlagEncoder flagEncoder = getEncodingManager().getEncoder(encoderName);
-					coreEdgeFilter.add(new MaximumSpeedCoreEdgeFilter(flagEncoder, maximumSpeedLowerBound));
-					if (flagEncoder.supports(TurnWeighting.class))
-						coreEdgeFilter.add(new TurnRestrictionsCoreEdgeFilter(flagEncoder, gs));
-					break;
-				}
-			}
-		}
-
-		/* End filter sequence initialization */
-
 		//Create the core
 		if(coreFactoryDecorator.isEnabled())
-			coreFactoryDecorator.createPreparations(gs, coreEdgeFilter);
+			coreFactoryDecorator.createPreparations(gs, processContext);
 		if (!isCorePrepared())
 			prepareCore();
 

--- a/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/CoreAlgoFactoryDecorator.java
+++ b/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/CoreAlgoFactoryDecorator.java
@@ -16,11 +16,19 @@ package org.heigit.ors.routing.graphhopper.extensions.core;
 import com.graphhopper.routing.RoutingAlgorithmFactory;
 import com.graphhopper.routing.RoutingAlgorithmFactoryDecorator;
 import com.graphhopper.routing.util.EdgeFilter;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.FlagEncoder;
 import com.graphhopper.routing.util.HintsMap;
 import com.graphhopper.routing.weighting.AbstractWeighting;
+import com.graphhopper.routing.weighting.TurnWeighting;
 import com.graphhopper.storage.*;
 import com.graphhopper.util.CmdArgs;
 import com.graphhopper.util.Helper;
+import org.heigit.ors.routing.RoutingProfileCategory;
+import org.heigit.ors.routing.graphhopper.extensions.GraphProcessContext;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.core.*;
+import org.heigit.ors.routing.graphhopper.extensions.flagencoders.FlagEncoderNames;
 import org.heigit.ors.routing.graphhopper.extensions.util.ORSParameters.Core;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -307,16 +315,17 @@ public class CoreAlgoFactoryDecorator implements RoutingAlgorithmFactoryDecorato
         }
     }
 
-    public void createPreparations(GraphHopperStorage ghStorage, EdgeFilter restrictionFilter) {
+    public void createPreparations(GraphHopperStorage ghStorage, GraphProcessContext processContext) {
         if (!isEnabled() || !preparations.isEmpty())
             return;
         if (!hasCHProfiles())
             throw new IllegalStateException("No profiles found");
 
         for (CHProfile chProfile : chProfiles) {
-            addPreparation(createCHPreparation(ghStorage, chProfile, restrictionFilter));
+            addPreparation(createCHPreparation(ghStorage, chProfile, createCoreEdgeFilter(ghStorage, processContext)));
         }
     }
+
     private PrepareCore createCHPreparation(GraphHopperStorage ghStorage, CHProfile chProfile, EdgeFilter restrictionFilter) {
         PrepareCore tmpPrepareCore = new PrepareCore(
                 new GHDirectory("", DAType.RAM_INT), ghStorage, ghStorage.getCHGraph(chProfile), restrictionFilter);
@@ -326,4 +335,49 @@ public class CoreAlgoFactoryDecorator implements RoutingAlgorithmFactoryDecorato
                 setLogMessages(preparationLogMessages);
         return tmpPrepareCore;
     }
+
+    private EdgeFilter createCoreEdgeFilter(GraphHopperStorage gs, GraphProcessContext processContext) {
+        EncodingManager encodingManager = gs.getEncodingManager();
+
+        int routingProfileCategory = RoutingProfileCategory.getFromEncoder(encodingManager);
+
+        /* Initialize edge filter sequence */
+        EdgeFilterSequence edgeFilterSequence = new EdgeFilterSequence();
+
+        /* Heavy vehicle filter */
+        if (encodingManager.hasEncoder(FlagEncoderNames.HEAVYVEHICLE)) {
+            edgeFilterSequence.add(new HeavyVehicleCoreEdgeFilter(gs));
+        }
+
+        /* Avoid features */
+        if ((routingProfileCategory & (RoutingProfileCategory.DRIVING | RoutingProfileCategory.CYCLING | RoutingProfileCategory.WALKING | RoutingProfileCategory.WHEELCHAIR)) != 0) {
+            edgeFilterSequence.add(new AvoidFeaturesCoreEdgeFilter(gs, routingProfileCategory));
+        }
+
+        /* Avoid borders */
+        if ((routingProfileCategory & (RoutingProfileCategory.DRIVING | RoutingProfileCategory.CYCLING)) != 0) {
+            edgeFilterSequence.add(new AvoidBordersCoreEdgeFilter(gs));
+        }
+
+        if (routingProfileCategory == RoutingProfileCategory.WHEELCHAIR) {
+            edgeFilterSequence.add(new WheelchairCoreEdgeFilter(gs));
+        }
+
+        /* Maximum speed & turn restrictions */
+        if ((routingProfileCategory & RoutingProfileCategory.DRIVING) !=0) {
+            String[] encoders = {FlagEncoderNames.CAR_ORS, FlagEncoderNames.HEAVYVEHICLE};
+            for (String encoderName: encoders) {
+                if (encodingManager.hasEncoder(encoderName)) {
+                    FlagEncoder flagEncoder = encodingManager.getEncoder(encoderName);
+                    edgeFilterSequence.add(new MaximumSpeedCoreEdgeFilter(flagEncoder, processContext.getMaximumSpeedLowerBound()));
+                    if (flagEncoder.supports(TurnWeighting.class))
+                        edgeFilterSequence.add(new TurnRestrictionsCoreEdgeFilter(flagEncoder, gs));
+                    break;
+                }
+            }
+        }
+
+        return edgeFilterSequence;
+    }
+
 }

--- a/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/PrepareCore.java
+++ b/openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/PrepareCore.java
@@ -41,7 +41,6 @@ public class PrepareCore extends AbstractAlgoPreparation implements RoutingAlgor
     private static final Logger LOGGER = Logger.getLogger(PrepareCore.class);
     private final CHProfile chProfile;
     private final PreparationWeighting prepareWeighting;
-    private final TraversalMode traversalMode;
     private final EdgeFilter restrictionFilter;
     private final GraphHopperStorage ghStorage;
     private final CHGraphImpl prepareGraph;
@@ -85,7 +84,6 @@ public class PrepareCore extends AbstractAlgoPreparation implements RoutingAlgor
         this.ghStorage = ghStorage;
         this.prepareGraph = (CHGraphImpl) chGraph;
         this.chProfile = chGraph.getCHProfile();
-        this.traversalMode = chProfile.getTraversalMode();
         this.weighting = chProfile.getWeighting();
         this.flagEncoder = weighting.getFlagEncoder();
         this.restrictionFilter = restrictionFilter;

